<a href="https://colab.research.google.com/github/antoniosolis4536-collab/MIAAD/blob/main/Actividad_Calificada_07_Datos_Duplicados_274706.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |ANTONIO ISMAEL SOLIS MARTINEZ|
| **Matrícula** |274706|

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [64]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [65]:
# Escribe tu código aquí
# Conteo de duplicados exactos usando keep
dup_first = hoteles.duplicated(keep='first').sum()
dup_last = hoteles.duplicated(keep='last').sum()
dup_false = hoteles.duplicated(keep=False).sum()

# Imprimir los resultados
print("Número de filas duplicadas (keep='first') no contempla el primero:\n", dup_first)
print("Número de filas duplicadas (keep='last') no contempla el ultimo:\n", dup_last)
print("Número de filas duplicadas (keep=False) contempla todas:\n", dup_false)

Número de filas duplicadas (keep='first') no contempla el primero:
 31994
Número de filas duplicadas (keep='last') no contempla el ultimo:
 31994
Número de filas duplicadas (keep=False) contempla todas:
 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [66]:
# Escribe tu código aquí
porcentaje_sobrante = dup_first / len(hoteles) * 100
print(f"Porcentaje de filas sobrantes: {porcentaje_sobrante:.2f}%")

# numero de grupos repetidos
conteo_combinaciones = hoteles.value_counts(dropna=False)
grupos_repetidos = conteo_combinaciones[conteo_combinaciones > 1].sum()
print(f"Número de grupos repetidos: {grupos_repetidos}")

Porcentaje de filas sobrantes: 26.80%
Número de grupos repetidos: 40165


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [67]:
# Escribe tu código aquí
hoteles_copia = hoteles.copy()
hoteles_copia['es_duplicado'] = hoteles_copia.duplicated(keep='first')

resumen_hoteles = hoteles_copia.groupby('hotel')['es_duplicado'].agg(['sum', 'count'])
resumen_hoteles['porcentaje'] = resumen_hoteles['sum'] / resumen_hoteles['count'] * 100
print(resumen_hoteles)


                sum  count  porcentaje
hotel                                 
City Hotel    25902  79330   32.650952
Resort Hotel   6092  40060   15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._ Me parece que si, porque hay muchas filas duplicadas como lo vimos el 26% y no le pega igual a los 2 hoteles porque los % son diferentes en cada hotel.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [68]:
# Escribe tu código aquí
columnas_vistas = ['hotel', 'lead_time', 'arrival_date_year',
                   'arrival_date_month', 'arrival_date_day_of_month',
                   'adults', 'country', 'market_segment', 'deposit_type',
                   'adr', 'reservation_status']
# Ordenaremos por las mismas columnas de vista para las copias de cada grupo esten juntas
duplicados_todos = hoteles[hoteles.duplicated(keep=False)]
duplicados_ordenados = duplicados_todos.sort_values(by=columnas_vistas)

duplicados_ordenados[columnas_vistas].head(12)


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40643,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40644,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40772,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40802,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40821,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
40838,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
76792,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76793,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76794,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
41067,City Hotel,0,2015,August,11,1,PRT,Corporate,No Deposit,88.0,Check-Out


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [69]:
# Escribe tu código aquí
conteo_grupos = hoteles.value_counts(dropna=False)

print("Tamaño de grupo mas frecuente (top 10):")
print(conteo_grupos[conteo_grupos > 1].value_counts().sort_index(ascending=False).head(10))

tamaño_maximo = conteo_grupos.max()
print(f"\nTamaño del grupo más grande: {tamaño_maximo}")
print("\nCombinación de valores del grupo más grande:")
print(conteo_grupos[conteo_grupos == tamaño_maximo])


Tamaño de grupo mas frecuente (top 10):
count
180    1
150    1
109    1
101    1
100    1
99     3
89     1
85     3
84     2
79     3
Name: count, dtype: int64

Tamaño del grupo más grande: 180

Combinación de valores del grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0    

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [70]:
# Escribe tu código aquí
mascara_dup = hoteles.duplicated(keep=False)
duplicados = hoteles[mascara_dup]

# Distribución de market_segment
print("Distribución de market_segment:")
print("Duplicados:\n", duplicados['market_segment'].value_counts(normalize=True))
print("\nCompletos:\n", hoteles['market_segment'].value_counts(normalize=True))

# Distribución de deposit_type
print("\nDistribución de deposit_type:")
print("Duplicados:\n", duplicados['deposit_type'].value_counts(normalize=True))
print("\nCompletos:\n", hoteles['deposit_type'].value_counts(normalize=True))

# Tasa de cancelación
print("\nTasa de cancelación:")
print("Duplicados:\n", duplicados['is_canceled'].mean())
print("Completos:\n", hoteles['is_canceled'].mean())

Distribución de market_segment:
Duplicados:
 market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64

Completos:
 market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64

Distribución de deposit_type:
Duplicados:
 deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64

Completos:
 deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Tasa de cancelación:
Duplicados:
 0.583866550479273
Completos:
 0.37041628277075134


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

_Escribe aquí tu respuesta._

a) Yo diria que son errores de registro, no reservas distintas. Es muy raro que dos personas distintas coincidan exactamente en todo, entonce lo mas logico es que se cargo duplicado.

b) diplicated() toma los NaN como si fueran iguales. Entonces si dos reservas no tienen agencia (es decir que sean NaN), las cuenta como coincidencia auqneu no sean el mismo caso.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [71]:
# Escribe tu código aquí
filas_antes = len(hoteles)
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)
filas_despues = len(hoteles_limpio)

print(f"Filas antes: {filas_antes}")
print(f"Filas después: {filas_despues}")
print(f"Filas eliminadas: {filas_antes - filas_despues}")


Filas antes: 119390
Filas después: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [72]:
# Escribe tu código aquí
resumen_original = hoteles.groupby('hotel').agg(
    reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    tarifa_promedio=('adr', 'mean')
).reset_index()

resumen_limpio = hoteles_limpio.groupby('hotel').agg(
    reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    tarifa_promedio=('adr', 'mean')
).reset_index()

print("Resumen original:")
print(resumen_original)

print("\nResumen limpio:")
print(resumen_limpio)

Resumen original:
          hotel  reservas  tasa_cancelacion  tarifa_promedio
0    City Hotel     79330          0.417270       105.304465
1  Resort Hotel     40060          0.277634        94.952930

Resumen limpio:
          hotel  reservas  tasa_cancelacion  tarifa_promedio
0    City Hotel     53428          0.300386       110.985944
1  Resort Hotel     33968          0.234809        99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [73]:
# Escribe tu código aquí
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)
comparacion = comparacion.sort_values(by='hotel').reset_index(drop=True)
comparacion

,hotel,reservas,tasa_cancelacion,tarifa_promedio,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._

a) La métrica que más se mueve es la que salga con mayor diferencia en la tabla comparacion que seria reservas y cancelacion, y el cambio es más grande en el hotel que tiene más duplicados (en este caso City Hotel). Tiene sentido con lo que vimos en la actividad 2, si los duplicados no están parejos entre segmentos, pues jalan más para un lado en el hotel donde ese segmento pesa más.

b) Si la misma reserva queda una vez en entrenamiento y otra en prueba, el modelo básicamente "ya vio la respuesta" antes de la evaluación. Eso hace que parezca más bueno de lo que realmente es.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [74]:
# Escribe tu código aquí
print("Duplicados exactos:")
print(hoteles_limpio.duplicated(keep=False).sum())
print("\nDuplicados por índice:")
print(hoteles_limpio.index.duplicated().sum())
print("\nIndice va de 0 a len-1:")
print(list(hoteles_limpio.index) == list(range(len(hoteles_limpio))))

Duplicados exactos:
0

Duplicados por índice:
0

Indice va de 0 a len-1:
True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [75]:
# Escribe tu código aquí
columnas_sin_fecha = hoteles_limpio.columns.drop('reservation_status_date')

dup_parcial = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep=False)
print("Filas sobrantes con este criterio:", hoteles_limpio[dup_parcial].duplicated(subset=columnas_sin_fecha).sum())

cols_vista = ['hotel', 'lead_time', 'arrival_date_year',
              'arrival_date_month', 'arrival_date_day_of_month',
              'adr', 'reservation_status', 'reservation_status_date']

orden_cols = [i for i in columnas_sin_fecha if i != 'reservation_status_date'] + ['reservation_status_date']
dup_parcial_orden = hoteles_limpio[dup_parcial].sort_values(by=orden_cols)
print(dup_parcial_orden[cols_vista].head(12))

print(hoteles_limpio[dup_parcial]['reservation_status'].value_counts())

Filas sobrantes con este criterio: 245
            hotel  lead_time  arrival_date_year arrival_date_month  \
35574  City Hotel          9               2015          September   
35572  City Hotel          9               2015          September   
51268  City Hotel         12               2017               July   
51267  City Hotel         12               2017               July   
54170  City Hotel         14               2015            October   
54474  City Hotel         14               2015            October   
39432  City Hotel         14               2016              April   
39425  City Hotel         14               2016              April   
40080  City Hotel         14               2016                May   
40083  City Hotel         14               2016                May   
45793  City Hotel         15               2016           December   
45797  City Hotel         15               2016           December   

       arrival_date_day_of_month     adr reservati

**Respuesta 4.2:**

_Escribe aquí tu respuesta._ Esto seguramente es la misma reserva guardada en distintas fechas de actualización de estado, no una reserva nueva. No las quitaría inmediatamente, porque esa fecha sí dice algo real (cuándo cambió el estado) al borrarlas perderia el historia de la informacion.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._


**Criterio:** consideré duplicado cuando coinciden las 32 columnas completas, sin excluir ninguna.

**Conservación:** usé keep='first'. Aquí no importa mucho cuál te quedes, porque si son idénticas en todo, da lo mismo cuál copia sea.

**Resultado:** Las cifras filas_antes = 119,390 y filas_despues = 87,396 y su diferencia = 31,994.

**Supuestos:** asumo que si coinciden en todo es error de captura, no que sea casualidad. No hay forma de estar 100% seguro porque no hay un ID de reserva. También hay que tener en cuenta que agent / company con NaN pudo haber inflado un poco el conteo.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

_Escribe aquí tu respuesta._

**1.** No la quitaría si lo que quiero es ver cuántas veces se mandó la misma solicitud al sistema entonces ahí el duplicado es justo el dato que me interesa, no un error a corregir.

**2.** Pediría un ID de reserva y la fecha exacta en que se creó el registro. Con eso sí podría estar seguro si es error o coincidencia real, ahorita solo estoy suponiendo.

**3.** Porque duplicated() solo cacha lo que es exactamente igual. Si hay espacios de más, mayúsculas distintas, o el mismo dato guardado en dos formatos diferentes, ahí no lo detecta aunque sea el mismo caso repetido.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.